analyzers/src/SonarAnalyzer.CSharp/Rules/LockedFieldShouldBeReadonly.cs

```diff
@@ -0,0 +1,79 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class LockedFieldShouldBeReadonly : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S2445";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, "Do not lock on {0}, use a readonly field instead.");
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(CheckLockStatement, SyntaxKind.LockStatement);
+
+    private static void CheckLockStatement(SonarSyntaxNodeReportingContext context)
+    {
+        var expression = ((LockStatementSyntax)context.Node).Expression?.RemoveParentheses();
+        if (IsCreation(expression))
+        {
+            ReportIssue("a new instance because is a no-op");
+        }
+        else
+        {
+            var lazySymbol = new Lazy<ISymbol>(() => context.SemanticModel.GetSymbolInfo(expression).Symbol);
+            if (IsOfTypeString(expression, lazySymbol))
+            {
+                ReportIssue("strings as they can be interned");
+            }
+            else if (expression is IdentifierNameSyntax && lazySymbol.Value is ILocalSymbol lockedSymbol)
+            {
+                ReportIssue($"local variable '{lockedSymbol.Name}'");
+            }
+            else if (FieldWritable(expression, lazySymbol) is { } lockedField)
+            {
+                ReportIssue($"writable field '{lockedField.Name}'");
+            }
+        }
+
+        void ReportIssue(string message) =>
+            context.ReportIssue(Diagnostic.Create(Rule, expression.GetLocation(), message));
+    }
+
+    private static bool IsCreation(ExpressionSyntax expression) =>
+        expression.IsAnyKind(
+            SyntaxKind.ObjectCreationExpression,
+            SyntaxKind.AnonymousObjectCreationExpression,
+            SyntaxKind.ArrayCreationExpression,
+            SyntaxKind.ImplicitArrayCreationExpression,
+            SyntaxKind.QueryExpression);
+
+    private static bool IsOfTypeString(ExpressionSyntax expression, Lazy<ISymbol> lazySymbol) =>
+        expression.IsAnyKind(SyntaxKind.StringLiteralExpression, SyntaxKind.InterpolatedStringExpression)
```

Comment:
Also `Utf8StringLiteralExpression`. You might need to add a member into `SyntaxKindEx` shim enum

Reply:
`SyntaxKindEx.Utf8StringLiteralExpression` it's there.

Reply:
Unless I am missing a referenced-type, string-like version of a UTF-8 string, it doesn't make sense to check against `Utf8StringLiteralExpression`.

`Utf8StringLiteralExpressionSyntax`, to my (very limited) knowledge, is only generated by `SyntaxToken`s in the form `"..."u8`. Those are just syntactic sugar for a Span of some sort, not for a `string`: a value type*, not a reference type!

As such, you can never do something like `lock("string literal"u8) { }`, pretty much in the same way you cannot do `lock(42) {}` or `lock(new AnyOtherStruct(...)) {}` etc...

I have already added test cases (resolving into compiler error CS0185) for some literals resolving into structs, such as `int` and value tuples. I can add more CS0185 errors for spans, but it seems not really useful to me.

* As for `ReadOnlySpan` being a `ref struct` - that changes stack allocation details, and regardless, it doesn't change whether it's a reference or a value type.

Reply:
What's the point of having `SyntaxKind.StringLiteralExpression` here then? Is it for `lock("hello")` that also doesn't make sense?
It should make sense and be consistent.

Reply:
The utf8 doens't compile

Reply:
As discussed offline, keeping `Utf8StringLiteralExpressionSyntax` out, since `lock("string"u8) { }` doesn't compile.